### PR TITLE
Redesign App Structure

### PR DESCRIPTION
**Pull request by akshay2905**
Redesigning of the app to generic components with can be used by different types of data. Though a lot has been changed, the app would require further restructuring using redux. The major changes include - separating of most of the business logic to single component, making the components generic, addition of Receipts Data, increasing the scope of hierarchy the app can handle.

### CODE REVIEW

File: src/components/controller/AppController.js [hunk 1]
```diff
@@ -1,49 +1,162 @@
...
     constructor(props) {
         super(props);
         this.state = { 
-            budgetAttr:"BE",
+            heirarchyLevel : this.props.params.heirarchy_level,
+            categorySlug : this.props.params.category,
+            recordSlug : this.props.params.record,
+            series : this.props.route.panelName,
             viewBy: "choropleth",
+            budgetAttr : "BE",
```
Comment (gggodhwani) [inline src/components/controller/AppController.js]: Can we make this generic too? Attribute1 or so?

File: src/components/controller/AppController.js [hunk 2]
```diff
@@ -1,49 +1,162 @@
...
+
+    setRecordSlug(){
+        this.setState({recordSlug:this.props.params.record});
+    }
+
+    setRecordData(){
+        let currentState = this.state;
+        if(this.props.params.heirarchy_level == 2 ){
+            let recordData = appController.dataseries.find(function(series){
+                return series.title_slug == currentState.series;
```
Comment (gggodhwani) [inline src/components/controller/AppController.js]: series.titleSlug?

File: src/components/controller/AppController.js [hunk 3]
```diff
@@ -1,49 +1,162 @@
...
+                }).data.find(function(category){
+                    return category.category_slug == currentState.categorySlug;
```
Comment (gggodhwani) [inline src/components/controller/AppController.js]: Same camel case?

File: src/components/views/leftsidebar/subcomponents/TabsPanel.js [hunk 4]
```diff
@@ -105,35 +130,45 @@ class PanelContent extends React.Component{
...
     this.state = {
     panelData : null, 
      activeTab: null
-    }
-    this.handleClick = this.handleClick.bind(this)
+    };
+    this.handleClick = this.handleClick.bind(this);
   }
   
+  componentWillMount(){
```
Comment (gggodhwani) [inline src/components/views/leftsidebar/subcomponents/TabsPanel.js]: Add a comment to explain what this function will be mounting and where?

Comment (akshay2905) [inline src/components/views/leftsidebar/subcomponents/TabsPanel.js]: This is one of the React lifecycle's methods which are used in the component lifecycle.
Additional Information - [componentWillMount()](https://facebook.github.io/react/docs/react-component.html#componentwillmount). What it is doing is setting states of different states using the props the component receives. 
We probably need to look and accept an extensive styling guide not just for code but for documentation too.

File: src/components/controller/AppController.js (hunk 2 above)
Comment (akshay2905) [inline src/components/controller/AppController.js]: series is the parameter for the function. titleSlug is the object's parameters.

File: src/components/controller/AppController.js (hunk 3 above)
Comment (akshay2905) [inline src/components/controller/AppController.js]: Camel Case has been used for React app variables. While the variable style for external variables(external data) used is different than the one used for app.

File: src/components/controller/AppController.js (hunk 1 above)
Comment (akshay2905) [inline src/components/controller/AppController.js]: This is part of the bigger issue we are facing. To what extent can we make our React Components flexible is a subjective matter, which depends on the release cycle. While rigid components are less generic, they are easier to build. While a flexible component is more extensible it requires more time, which would affect the release cycle.

**Pull request merged**